bin/cheribsdtest/cheribsdtest.c

```diff
@@ -623,6 +628,103 @@ cheribsdtest_run_test_name(const char *name)
 	errx(EX_USAGE, "unknown test: %s", name);
 }
 
+static void
+cheribsdtest_run_child(struct cheri_test *ctp)
+{
+	if (ctp->ct_child_func == NULL)
+		errx(EX_SOFTWARE, "%s has no child function", ctp->ct_name);
+	ctp->ct_child_func();
+	errx(EX_SOFTWARE, "%s child function returned", ctp->ct_name);
+}
+
+static void
+cheribsdtest_run_child_name(const char *name)
+{
+	struct cheri_test **ctpp, *ctp;
+
+	SET_FOREACH(ctpp, cheri_tests_set) {
+		ctp = *ctpp;
+		if (strcmp(name, ctp->ct_name) == 0) {
+			cheribsdtest_run_child(ctp);
+		}
+	}
+	errx(EX_USAGE, "unknown test: %s", name);
+}
+
+static char **
+mk_exec_args(const struct cheri_test *ctp)
+{
+	char *execpath;
+	char const **exec_args;
+	int argc = 0, error;
+
+	execpath = malloc(MAXPATHLEN);
+	if (execpath == NULL)
+		err(EX_OSERR, "malloc");
+	exec_args = calloc(5, sizeof(*exec_args));
+	if (exec_args == NULL)
+		err(EX_OSERR, "calloc");
+
+	/*
+	 * XXX: This won't work for direct exec as an rtld argument.
+	 * (e.g., /libexec/ld-elf.so.1 /bin/cheribsdtest-purecap-dynamic)
+	 * If this becomes an issue we could alter rtld to update
+	 * AT_EXECPATH or add some sort of coexeve_self(3) implemented
+	 * by rtld for dynamic binaries and libc for static.
+	 */
+	error = elf_aux_info(AT_EXECPATH, execpath, MAXPATHLEN);
+	if (error != 0)
+		errx(EX_OSERR, "elf_aux_info: %s", strerror(error));
+	exec_args[argc++] = execpath;
+	exec_args[argc++] = "-E";
+	if (coredump_enabled)
+		exec_args[argc++] = "-c";
+	if (verbose)
+		exec_args[argc++] = "-v";
+	exec_args[argc++] = ctp->ct_name;
+	exec_args[argc++] = NULL;
+
+	return (__DECONST(char **, exec_args));
+}
+
+pid_t
+cheribsdtest_spawn_child(int mode)
+{
+        char **exec_args;
+        int error;
+	pid_t pid;
+
+	switch (mode) {
+		case SC_MODE_POSIX_SPAWN:
+			exec_args = mk_exec_args(running_test);
+			error = posix_spawn(&pid, exec_args[0], 
+				NULL, NULL, exec_args, environ);
+                        if (error != 0)
+                                err(EX_OSERR, "%s: posix_spawn", __func__);
+			break;
+		case SC_MODE_FORK:
+			pid = fork();
+			break;
+		case SC_MODE_RFORK:
+			pid = rfork(RFPROC);
+			break;
+		case SC_MODE_VFORK:
+			pid = vfork();
+			break;
+		default:
```

Comment:
```suggestion
		default:
			errno = EDOOFUS;
```
Indicate that this is a programming error.